### DISCUSSION TITLE
Microsoft.Extensions.Http.Polly support for v8

### DISCUSSION THREAD

**Discussion by jonsagara** [App-vNext/Polly#1657] [Q&A] [answered]
Congrats on the v8 release! 😎

Do you have any visibility into when `Microsoft.Extensions.Http.Polly` will support Polly v8?

Thanks!

**Comment by martincostello** [comment, marked as answer]
As far as I'm aware, Microsoft have no plans to update Microsoft.Extensions.Http.Polly to natively use the new v8 APIs as the new libraries being released as part of .NET 8 from the [dotnet/extensions](https://github.com/dotnet/extensions) repository (such as Microsoft.Extensions.Resilience and Microsoft.Extensions.Http.Resilience) will build upon our new API surface and supersede the previous extension which comes from the [dotnet/aspnetcore](https://github.com/dotnet/aspnetcore/tree/main/src/HttpClientFactory/Polly/src) repository.

**Reply by jonsagara** [reply]
Got it. Thanks!

**Reply by IEvangelist** [reply]
See this article that was literally just released 1o days ago: https://learn.microsoft.com/dotnet/core/resilience. I believe you'll have until .NET 9 to migrate over as not everything is in place yet to make that easy/possible.

**Comment by joelhulen** [comment]
I'm planning on testing `Microsoft.Extensions.Http.Resilience` today and will report back if any issues.

**Comment by glen-84** [comment]
I'm getting really confused by all of these changes. I just want to add a registry to the service collection, and then associate a pipeline in the registry with an HTTP client.

- For the former, I would have expected `AddResiliencePipelineRegistry` to take a `ResiliencePipelineRegistry`, but it doesn't. Not sure how that works.
- For the latter, I would have expected something like `AddResilienceHandlerFromRegistry("MyPipeline")`, but that doesn't exist.

I'm also not sure how to do the equivalent of `HandleTransientHttpError`, perhaps some combination of `ShouldHandle` and `HttpClientResiliencePredicates.IsTransientHttpException`?

**Reply by martincostello** [reply]
For your first point, this sounds like a feature request where instead of you asking for one to be added you want to make it yourself. Please open a new issue with information about your use case.

For the second point, that functionality sounds like something that would be part of [dotnet/extensions](https://github.com/dotnet/extensions) and the Microsoft.Extensions.Http.Resilience library. If my understanding is correct, please file an issue there for the .NET team to look at.

**Reply by glen-84** [reply]
> For your first point, this sounds like a feature request where instead of you asking for one to be added you want to make it yourself. Please open a new issue with information about your use case.

I didn't realize that calling `AddResiliencePipeline` would also add the pipeline to the registry, so this is fine now.

> For the second point, that functionality sounds like something that would be part of dotnet/extensions and the Microsoft.Extensions.Http.Resilience library. If my understanding is correct, please file an issue there for the .NET team to look at.

I've opened https://github.com/dotnet/extensions/issues/4666.

Thanks for your help.

**Comment by anktsrkr** [comment]
I tried to migrate the existing MS extension to make it work with Polly v8.  You can find [nuget](https://www.nuget.org/packages/Sundry.Extensions.Http.Polly) here. It was easy to migrate and looks like every scenario is working as expected. I understand all these changes are breaking, apart from that just wondering if I missed something for which MS reimplementing from scratch and had a quick look seems quite complex than the previous implementation.

**Reply by martincostello** [reply]
The re-engineering of Polly was driven primarily to improve performance and reduce allocations, which the previous API didn't lend itself to due iterating on due to the fundamentals of how it was originally designed.

We did consider a compatibility layer to redirect the old APIs into the new Polly.Core API, but when we investigated it just made the old API slower due to the overhead of having to bridge between the two quite different API surfaces.

**Reply by anktsrkr** [reply]
@martincostello after you pointed out the motivation of new Microsoft's extension, I have started looking into it. Seems like we don't have much documentation yet( I may miss as well :-)), so I have started it which is specific to httpclient factory. If you find some time, have a look here - https://anktsrkr.github.io/post/implementing-retry-strategy-using-httpclientfactory-with-pollyv8-and-.net-8/ and let me know if you have any inputs. Thanks.

**Reply by martincostello** [reply]
@IEvangelist Could you point us to the documentation you've been working on for the new resilience libraries?

**Reply by IEvangelist** [reply]
Yes sir:

- [Introduction to resilient app development](https://learn.microsoft.com/dotnet/core/resilience)
- [Build resilient HTTP apps: Key development patterns](https://learn.microsoft.com/dotnet/core/resilience/http-resilience)

**Reply by martintmk** [reply]
The blog post introducing the new library is now also published:

https://devblogs.microsoft.com/dotnet/building-resilient-cloud-services-with-dotnet-8/

**Comment by joelhulen** [comment]
Great blog post, Martin! Thank you for sharing :)
